### ISSUE TITLE
Adding document counts to the query API

### ISSUE THREAD

**Issue by hyzx86** [OrchardCMS/OrchardCore#16578]
### Is your feature request related to a problem? Please describe.

Currently, some OC query APIs such as `LuceneQuery` , `SqlQuery` and `graphQL Query` do not yet support statistical aggregation.

### Describe the solution you'd like

1. Lucene Query Elastic Search Query 
These api's make it easy to implement statistics. We just need to refactor the graphQL provider associated with them. 2. SqlQuey In SqlQuery we should allow users to write SQL for statistical queries.

2. SqlQuey 
In SqlQuery, we should allow the user to write SQL for statistical queries, perhaps with additional script fields, or some other better way of accomplishing this, and the current API doesn't seem to support it.

4. the script function `executeQuery`, whose return value is currently always a collection.

### Describe alternatives you've considered

If we can't implement these features anytime soon, I suggest that before the[ 2.0 release](https://github.com/OrchardCMS/OrchardCore/issues/16428), we first return `IQueryResult` objects in all locations where we need to call `queryManager.ExecuteQueryAsync` to return data, instead of just returning the collection.

**Comment by hishamco** [comment]
Can you propose a PR?

**Comment by hyzx86** [comment]
> Can you propose a PR?

I'd like to discuss the feasibility of these features first, since I put up a related PR about many years ago, which seemed to be unaccepted 

https://github.com/OrchardCMS/OrchardCore/pull/11159

https://github.com/OrchardCMS/OrchardCore/issues/5693

**Comment by MikeAlhayek** [comment]
I think having a count is useful specially if you need to create a pager. Is there a reason why this was rejected? Does all the search providers gives you a way to get a count without returning the list of contentItemId?

**Comment by hyzx86** [comment]
> Does all the search providers gives you a way to get a count without returning the list of contentItemId? 

From my understanding of OC, Lucene already includes totals, we just don't open them up, SqlQuery definitely doesn't support them, if we need to support them, my suggestion is to provide a switch for the query to query for stats, and when the user opens it up the page will add a stats count of the dynamic `grpahql fields` generated by the SQL based on the type of the content. grpahql fields` should be supported, but obviously needs to be restructured or, as mentioned in https://github.com/OrchardCMS/OrchardCore/issues/5693, use `Connections`.

**Comment by sebastienros** [comment]
1. I believe it shouldn't be an issue to expose and extra field for the count, but this should be opt-in, existing queries should not get the count by default or this could impact the perf. It has to be on-demand.
2. Can't you already create SQL queries that return any kind of data?
3. We could have a new `executeScalarQuery` to return an object

**Comment by sebastienros** [comment]
Should these be 3 distinct issues?

**Comment by hyzx86** [comment]
> Should these be 3 distinct issues?



> 1. I believe it shouldn't be an issue to expose and extra field for the count, but this should be opt-in, existing queries should not get the count by default or this could impact the perf. It has to be on-demand.我相信公开计数的额外字段应该不是问题，但这应该是选择加入的，现有查询默认不应获取计数，否则这可能会影响性能。它必须是按需的。

Yes I couldn't agree more, in `graphQL field provider` we can check if the user needs the `total` field to do a statistic query.

> 2. Can't you already create SQL queries that return any kind of data?难道您已经创建了返回任何类型数据的 SQL 查询吗？

Not at the moment, but I made a Jint-based `Script Query` that can return objects, or data in any format.

> 3. We could have a new `executeScalarQuery` to return an object我们可以有一个新的 `executeScalarQuery` 来返回一个对象

Yes, I added an `executeObjectQuery` to support `ScriptQuery` calls.

**Comment by hyzx86** [comment]
> Should these be 3 distinct issues?这应该是 3 个不同的问题吗？

The main point of bringing up this issue is to suggest that we check where we are calling queries that return collections instead of objects for subsequent extensions. 

Since it's the point in time when 2.0 is ready for release, and we've already done a lot of disruptive upgrades in OC 2.0, 
I'd suggest adding these tweaks to the 2.0 release schedule now

**Comment by PBMikeW** [comment]
I had to write a custom controller to get the total results back (needed for paging!) Elastic and Lucene already include a total count in LuceneQueryResults and ElasticQueryResults. So the total on those is effectively free, we could include an optional total on SQL results. 

Most document databases are not that big, however mine has over 400k and returning a count on all published documents takes 900ms, and all of my biggest contenttype is 400ms. So expensive but I would never use SQL for those. 

`public class LuceneQueryResults : IQueryResults
{
    public IEnumerable<object> Items { get; set; }

    public int Count { get; set; }
}`

**Comment by Skrypt** [comment]
![image](https://github.com/user-attachments/assets/91a0f6d5-f5e1-4031-b8f9-322a4ffbd70b)

Here is how I fixed your code example 😉

**Comment by yqzhen1990** [comment]
> ![image](https://private-user-images.githubusercontent.com/3228637/377224575-91a0f6d5-f5e1-4031-b8f9-322a4ffbd70b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwMTY5OTcsIm5iZiI6MTczNTAxNjY5NywicGF0aCI6Ii8zMjI4NjM3LzM3NzIyNDU3NS05MWEwZjZkNS1mNWUxLTQwMzEtYjhmOS0zMjJhNGZmYmQ3MGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMjRUMDUwNDU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjlkZDIwOGIyZjE4YTRhMjM5YmJkY2VjYTEzMGU1OTAxZTJhMTRmNzg2NWY4M2JmODg4YWMzNzYwMDkwMDZjNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Y17YRrV8PMzcgT5i_nf4XSgDdTP9w5yllWZjygAVZio)
> 
> Here is how I fixed your code example 😉

![image](https://github.com/user-attachments/assets/6a41c46b-d563-4bda-8e32-22ad4de649c6)

If I want to get Count in the liquid template for use in paging, how can I do it?
I noticed that only result.Items is returned here:
![image](https://github.com/user-attachments/assets/27d68d05-bd04-4e77-a006-40116ab1f09e)
